sdk/src/dlob/DLOB.ts

```diff
@@ -0,0 +1,863 @@
+				const userAccountPublicKey = programAccount.publicKey;
+
+				for (const order of userAccount.orders) {
+					this.insertOrder(order, userAccountPublicKey);
+				}
+			}
+		}
+
+		this.initialized = true;
+		return true;
+	}
+
+	public insertOrder(
+		order: Order,
+		userAccount: PublicKey,
+		onInsert?: OrderBookCallback
+	): void {
+		if (isVariant(order.status, 'init')) {
+			return;
+		}
+
+		const marketType = getVariant(order.marketType) as MarketTypeStr;
+
+		if (isVariant(order.status, 'open')) {
+			this.openOrders
+				.get(marketType)
+				.add(getOrderSignature(order.orderId, userAccount));
+		}
+		this.getListForOrder(order).insert(
+			order,
+			marketType,
+			this.marketIndexToAccount
+				.get(marketType)
+				.get(order.marketIndex.toNumber()),
+			userAccount
+		);
+
+		if (onInsert) {
+			onInsert();
+		}
+	}
+
+	public trigger(
+		order: Order,
+		userAccount: PublicKey,
+		onTrigger?: OrderBookCallback
+	): void {
+		if (isVariant(order.status, 'init')) {
+			return;
+		}
+
+		const marketType = getVariant(order.marketType) as MarketTypeStr;
+
+		const triggerList = this.orderLists
+			.get(marketType)
+			.get(order.marketIndex.toNumber()).trigger[
+			isVariant(order.triggerCondition, 'above') ? 'above' : 'below'
+		];
+		triggerList.remove(order, userAccount);
+
+		this.getListForOrder(order).insert(
+			order,
+			marketType,
+			this.marketIndexToAccount
+				.get(marketType)
+				.get(order.marketIndex.toNumber()),
+			userAccount
+		);
+		if (onTrigger) {
+			onTrigger();
+		}
+	}
+
+	public getListForOrder(order: Order): NodeList<any> {
+		const isInactiveTriggerOrder =
+			isOneOfVariant(order.orderType, ['triggerMarket', 'triggerLimit']) &&
+			!order.triggered;
+
+		let type: DLOBNodeType;
+		if (isInactiveTriggerOrder) {
+			type = 'trigger';
+		} else if (isOneOfVariant(order.orderType, ['market', 'triggerMarket'])) {
+			type = 'market';
+		} else if (order.oraclePriceOffset.gt(ZERO)) {
+			type = 'floatingLimit';
+		} else {
+			type = 'limit';
+		}
+
+		let subType: string;
+		if (isInactiveTriggerOrder) {
+			subType = isVariant(order.triggerCondition, 'above') ? 'above' : 'below';
+		} else {
+			subType = isVariant(order.direction, 'long') ? 'bid' : 'ask';
+		}
+
+		const marketType = getVariant(order.marketType) as MarketTypeStr;
+		return this.orderLists.get(marketType).get(order.marketIndex.toNumber())[
+			type
+		][subType];
+	}
+
+	public findNodesToFill(
+		marketIndex: BN,
+		vBid: BN | undefined,
+		vAsk: BN | undefined,
+		slot: number,
+		marketType: MarketType,
+		oraclePriceData?: OraclePriceData
+	): NodeToFill[] {
+		// Find all the crossing nodes
+		const crossingNodesToFill: Array<NodeToFill> = this.findCrossingNodesToFill(
+			marketIndex,
+			vBid,
+			vAsk,
+			slot,
+			oraclePriceData
+		);
+
+		// TODO: verify that crossing nodes indeed include all market nodes? ok it's not, orders will be in one but not thet other zzz
```

Comment:
i think we just wanna make sure that market orders that are due to expire lead to a fill order ix